src/wmtk/TriMesh.cpp

```diff
@@ -18,6 +19,60 @@ TriMesh::TriMesh(TriMesh&& o) = default;
 TriMesh& TriMesh::operator=(const TriMesh& o) = default;
 TriMesh& TriMesh::operator=(TriMesh&& o) = default;
 
+Tuple TriMesh::split_edge(const Tuple& t)
+{
+    // record the deleted simplices topology attributes
+    TriMesh::TriMeshOperationExecutor executor(*this, t);
+    return executor.split_edge();
+}
+
+Tuple TriMesh::collapse_edge(const Tuple& t)
+{
+    // TODO: check link_cond before collapse
+    TriMeshOperationExecutor executor(*this, t);
+
+    // create new vertex
+    std::vector<long> new_vids = executor.request_simplex_indices(PrimitiveType::Vertex, 1);
+    assert(new_vids.size() == 1);
+    const long new_vid = new_vids[0];
+
+    // get faces in open_star(B)
+    auto star_A_f = SimplicialComplex::open_star(PrimitiveType::Vertex, t, *this)
+                        .get_simplices(PrimitiveType::Face);
+    auto star_B_f = SimplicialComplex::open_star(
+                        PrimitiveType::Vertex,
+                        switch_tuple(t, PrimitiveType::Vertex),
+                        *this)
+                        .get_simplices(PrimitiveType::Face);
+    std::vector<long> faces_to_change_fv;
+    for (const Simplex& simplex : star_B_f) {
```

Comment:
this code is older than the PR, but it breaks the semantics of a Simplex pretty harshly (isn't the point of a Simplex to lock down the interpretation of a Tuple to a single type? this is dynamically casting all simplices to an arbitrary face in its co-boundary.

this sort of logic makes me sorta wish SimplicialComplex held a vector of vector of Tuple rather than a single vector of Simplex objects :shrug:

Reply:
I think there is some misunderstanding. the `.get_simplices(...)` does not cast simplices into faces. It just extracts all faces that are inside a simplicial complex. I think the behavior is exactly what you want it to be.